### ISSUE TITLE
Segfault in cache::get/has

### ISSUE THREAD

**Issue by springmeyer** [mapbox/carmen-cache#81]
With https://github.com/mapbox/carmen/commit/c16b49098463065a when running: `bench/phrasematch.js` the process segfaults.

I've been debugging this for many hours without finding a clear culprit, so I'll document what I'm seeing and thinking:

### What I'm seeing:

 - I can replicate with every carmen-cache version I've tried so far: `master`, `v0.16.3`, and `v0.15.0`, `v0.14.0`, and `v0.12.1`
 - The crash alternates between `cache::_get` and `cache::has`
 - The crash does not happen if the dawg cache is not dumped and reloaded like done in https://github.com/mapbox/carmen/commit/4a1b748360ec984f6e79c080a85a7d4b505adc77.
 - The crash is in the `std::string` constructor
 - It happens because a null ptr is passed to the `std::string` constructor and `std::strlen` tries to compute its length. This can be reduced to:
  ```cpp
    const char * ptr = NULL;
    std::string shard = ptr; // will segfault on osx (libc++) and throw 'std::logic_error' on linux (libstdc++)
  ```
 - Putting a `console.log` before the call to `_get` or `has` in JS slows down the benchmark and 99% of the time avoids the crash.
 - One time the process did still crash after putting a `console.log` in to count how many times `phrasematch` had been called in the benchmark. This helped me see that `phrasematch` had run successfully ~ 11,000 before crashing.
 - Protecting against the crash in `cache::has` [here](https://github.com/mapbox/carmen-cache/blob/70cafc41c358d1510a90c6cb04e44f7ed7af3fa0/src/binding.cpp#L699) by throwing an error on a empty string argument `var->ToString().IsEmpty()` moves the crash predictably to `cache::get`
 - Protecting against the crash in both `cache::has` and `cache::_get` (by throwing a JS error on an empty string) leads to no segfault AND also no JS errors being thrown. Crazy, right?
 - So, protecting against creating a `std::string` from a NULL appears to stop there ever being a NULL???

### What I'm thinking:

 - I think the reason this happens with a compact `dawg` cache is because the program runs faster and the problem condition happens predictably.
 - I've noticed that if I put a print statement (like a console log before `_get` or `has`) then the problem goes away, another indication that the program needs to run fast to hit the problem

I think there are multiple bugs in play. One bug is that we don't protect from invalid/null strings from being sent to `std::string`. However, while we should protect from this, my sense is that a null ptr should not likely happen under normal conditions. So I suspect memory corruption or a race condition somewhere else in carmen cache loading. This then creates another condition where other bugs like the lack of NULL checking, manifest.

**Comment by apendleton** [comment]
If you had just mentioned that there was a race condition somewhere in carmen-cache leading to invalid reads, I would have guessed that it was in `coalesce`, which calls `__get` (or `__getTruncated`) in a thread other than the main thread, while the main thread might potentially be mutating the cache structures `__get` would access, either with calls to `unload`, or with calls to `loadSync` which cause cache eviction (as the cache has a limited size). In particular, I think there's a race condition opportunity [here](https://github.com/mapbox/carmen-cache/blob/master/src/binding.cpp#L104-L106) between the cache membership check and the cache fetch in `_get`, where if `_get` is called from `coalesce` it might be running in the thread pool, and in the intervening time the entry could get evicted or unloaded from the cache in the main thread.

But that's not what's going on here, if the crash is in the `has` and `_get` NAN_METHODs. `has`, `_get`, `loadSync`, and 'unload' should all only get called from JS, so they should all be running on the main thread, and the latter two are the only things that might invalidate cache contents as far as I know, so I'm pretty confused as to how mutation could be occurring during a call to `has` or `_get`, as I don't think it should be possible for any of the mutating functions to be called at the same time as those two. It's super weird.

**Comment by springmeyer** [comment]
Thank you for the thoughts! Right in this case neither of the asynchronous functions are being invoked (coalesce or merge). But i'll be digging into this more late tonight or tomorrow am to see if i can surface any more clues.

**Comment by springmeyer** [comment]
Recapping: because this is so mysterious I suspect memory corruption is causing it (so the bug might be elsewhere and only manifesting in get/has calls because they are called so often and will fall prey to memory being messed up elsewhere). Therefore, my game plan is:

 - Solve all the overflow issues flags by the sanitizers first: #75, #79 and #83
 - Instrument libc++ with the address sanitizer
 - Then revisit this test to try to replicate again.

**Comment by springmeyer** [comment]
Recapping here:

> Solve all the overflow issues flags by the sanitizers first: #75, #79 and #83

This was done in #86

> Instrument libc++ with the address sanitizer

Still not go around to this. Ticketed at https://github.com/mapbox/mason/issues/411

> Then revisit this test to try to replicate again.

No plans to at this point. Going to hope that the actual bug was fixed in #86.
